### ISSUE TITLE
import resolution incorrect with same-named file and external module

### ISSUE THREAD

**Issue by silverwind** [PyCQA/pylint#5151]
### Bug description

When a python file and a external module have the same name, python resolves a `import` for that name to the external module, but pylint resolves it to the current file which causes false positives in rules `import-self`, `no-name-in-module`, `no-member` and more module-related rules.

To reproduce:

- Install `redis` pypi module using your favourite package manager
- Create a file `redis.py` with content `import redis`
- Lint the file using `pylint redis.py`

### Pylint output

```shell
************* Module redis
redis.py:1:0: W0406: Module import itself (import-self)
```

### Expected behavior

No error

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.2
Python 3.9.7 (default, Sep  3 2021, 12:45:31)
[Clang 12.0.0 (clang-1200.0.32.29)]
```

### OS / Environment

macOS 11

### Additional dependencies

redis==3.2.0

**Comment by timmartin** [comment]
I think if you create an file called `redis.py` in the current directory, with the content:

    import redis

and that file is not part of a package then that file _is_ importing itself, because the directory of the currently-executing script will always be first on `sys.path`, and so the import will find the local file `redis.py` before it searches libraries.

Where there might be an issue is with a file called `redis.py` within a package. In that case, the directory containing the `redis.py` will no longer be on that path. But in that case, I can't reproduce a case where pylint will complain about it importing itself. If I have the file be in a package (just add an `__init__.py` to the directory) then pylint stops complaining about importing self.

I suppose if the file in question is imported from another Python file, and is not in a package but is imported because its directory has been added to `sys.path` after the standard import locations, then pylint will report it as import self and Python will not, although I haven't tested this. I don't think pylint can know about runtime modification to `sys.path`.

Is there a more specific case that we should be investigating here?

**Comment by jacobtylerwalls** [comment]
I suggest closing this in favor of #7289 and retitling it -- that one focuses on the case where we have `redis` or `gzip` instead of `redis.py`.

**Comment by DanielNoord** [comment]
Sorry, I missed the investigation by Tim. That said, I can still get an incorrect `import-self` to trigger when I create a `pylint.py` file in the root of the `pylint` project with a `import pylint` line.

https://github.com/PyCQA/astroid/pull/1747 fixes this.

**Comment by jacobtylerwalls** [comment]
Thanks @DanielNoord. Can you include a snippet from a python session that verifies that python is not importing self? E.g. `import pylint; print(pylint)` and show the output?

**Comment by DanielNoord** [comment]
```python
❯ cat pylint.py
import pylint

print(pylint.modify_sys_path)
❯ python pylint.py
<function modify_sys_path at 0x101dde050>
```

**Comment by DanielNoord** [comment]
Oh wait, https://github.com/PyCQA/astroid/pull/1747 does fix this. As the `no-member` warnings go away. As I explained in https://github.com/PyCQA/pylint/issues/7289#issuecomment-1221627013 for the `import-self` to go away we need to actually change the `import-self` check.

But the successful removal of `no-member` warnings on `pylint.modify_sys_path` shows that this works.

**Comment by jacobtylerwalls** [comment]
I see something different:
```
$ cat gzip.py
import gzip
print(gzip)

$ python3 gzip.py
<module 'gzip' from '/Users/jwalls/somefolder/gzip.py'>
<module 'gzip' from '/Users/jwalls/somefolder/gzip.py'>
```

**Comment by DanielNoord** [comment]
Yeah, this is different than the original issue but I think there is still an issue here.
If there is a folder `pylint` which is a normal python package and there is a `pylint.py` file that imports `pylint` then the `pylint` directory gets precedence. However, due to how we cache modules this doesn't work in `astroid`. https://github.com/PyCQA/astroid/pull/1747 does fix this.

**Comment by DanielNoord** [comment]
You can also recreate this with:
```console
cd /tmp
mkdir pylint
cd pylint
echo -e "X = 1" > __init__.py
cd ..
echo -e "import pylint\nprint(pylint.X)" > pylint.py
```

Without the fix in `astroid` this will warn about no-member, while Python executes this fine.

**Comment by DanielNoord** [comment]
Proposal, let https://github.com/PyCQA/astroid/pull/1747 fix this issue as this one was originally about import order resolution which I clearly what we are fixing there.

Rescope #7289 to fix `import-self` after we have access to that `astroid` update so that we can use the `Module` object instead of doing a string comparison for `import-self`.

@jacobtylerwalls Does that sound right?

**Comment by DanielNoord** [comment]
This was fixed in https://github.com/PyCQA/astroid/pull/1747. We don't need regression tests in `pylint` but this should be closed with the bump to astroid `2.13` in `pylint` at some point.

**Comment by jacobtylerwalls** [comment]
This was fixed with pylint 2.16's upgrade to astroid 2.13; please reopen if still an issue.

**Comment by dubiousjim** [comment]
Problem persists: see [this comment on #3665](https://github.com/pylint-dev/pylint/issues/3665#issuecomment-2449241821).
